### PR TITLE
Replace use of recursive-readdir package with local implementation

### PR DESCRIPTION
**Pull request by mattbjordan**
Fixes https://github.com/iTwin/itwinjs-core/issues/4524

[`recursive-readdir`](https://github.com/jergason/recursive-readdir) needs to update it's `minimatch` dependency version to fix a security vulnerability issue. Since the last update to this package was 3 years ago and [there has been an outstanding PR to upgrade the minimatch version since May](https://github.com/jergason/recursive-readdir/pull/85) with no response from the maintainer...

I adapted `recursive-readdir`'s code (<100 lines) into a single file in our build tools, removed the dependency on `minimatch` from it since it is not needed by us, and accredited the author and licensing accordingly. **Please let me know if you think the accreditation needs modification.**

### CODE REVIEW

**Review by raplemie — APPROVED** [review]
Will this be ported to 3.4.x ?

File: tools/build/scripts/utils/recursiveReaddir.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * recursive-readdir
+ * By Jamison Dance / GitHub User jergason
+ * Source: https://github.com/jergason/recursive-readdir
+ *
+ * Licensed under the MIT License.
+ * See tools\build\ThirdPartyNotices.md for full license.
+ */
+
+var fs = require("fs");
```
Comment (kabentley) [inline tools/build/scripts/utils/recursiveReaddir.js:12]: change all use of `var` to `const` or `let`.

**Review by kabentley — CHANGES_REQUESTED** [review]

File: tools/build/ThirdPartyNotices.md [hunk 2]
```diff
@@ -13,3 +13,29 @@ Permission is hereby granted, free of charge, to any person obtaining a copy of
 The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+## [recursive-readdir](https://github.com/jergason/recursive-readdir)
+
+*There is no attributed copyright year or copyright holders mentioned, but for clarity: the GitHub user who published the repository is 'jergason' who publishes their name as Jamison Dance. The following text is copied verbatim from [recursive-readdir](https://github.com/jergason/recursive-readdir).*
```
[the remark this answers is not in the document]
Comment (mattbjordan) [reply tools/build/ThirdPartyNotices.md]: Okay, sounds good to me. The fact that the license text was missing their name made me unsure.

File: tools/build/ThirdPartyNotices.md [hunk 3]
```diff
@@ -13,3 +13,29 @@ Permission is hereby granted, free of charge, to any person obtaining a copy of
 The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+## [recursive-readdir](https://github.com/jergason/recursive-readdir)
+
+*The following text is copied verbatim from recursive-readdir.*
```
Comment (kabentley) [inline tools/build/ThirdPartyNotices.md:19]: this line isn't necessary

**Comment by aruniverse** [comment]
> Will this be ported to 3.4.x ?

yes it will.

@mergifyio backport release/3.4.x

File: tools/build/ThirdPartyNotices.md [hunk 4]
```diff
@@ -13,3 +13,27 @@ Permission is hereby granted, free of charge, to any person obtaining a copy of
 The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+## [recursive-readdir](https://github.com/jergason/recursive-readdir)
+
+The MIT License (MIT)
+
+Copyright (c) \<year> \<copyright holders>
```
Comment (aruniverse) [inline tools/build/ThirdPartyNotices.md:21]: is this right? @wgoehrig

**Review by kabentley — APPROVED** [review]

File: tools/build/scripts/extract.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ const argv = require("yargs").argv;
 const path = require("path");
 const paths = require("./config/paths");
 const fs = require("fs-extra");
-const readDirectory = require("recursive-readdir");
+const readDirectory = require("./utils/recursiveReaddir");
```
Comment (aruniverse) [inline tools/build/scripts/extract.js:11]: can you verify the `extract-api` script is working as we expect it to? you should be able to delete the output for some of the existing `extract-api` output, and then rerun the script with your changes from this pr and there should be no diff between the 2.

Comment (mattbjordan) [reply tools/build/scripts/extract.js:11]: Already done. I both deleted all *.api.md and ran extract-api before and after the changes. There's no differences in the output.

**Comment by aruniverse** [comment]
@Mergifyio backport release/3.4.x

**Review by aruniverse — APPROVED** [review]

File: tools/build/ThirdPartyNotices.md (hunk 4 above)
Comment (kabentley) [reply tools/build/ThirdPartyNotices.md:21]: @aruniverse if the question is "do we have to reproduce their license file verbatim even if it is silly?" The answer is in the notice itself. It says that 

```txt
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
```
We must just copy whatever it says without changes to comply.

**Review by wgoehrig — APPROVED** [review]

**Comment by pmconne** [comment]
@mattbjordan nvm.

**Pull request merged**